pkg/protocol/stan/receiver.go

```diff
@@ -0,0 +1,207 @@
+	"io"
+)
+
+type msgErr struct {
+	msg binding.Message
+	err error
+}
+
+// Receiver allows stan Subscriptions to receive messages synchronously rather than via a callback handler
+type Receiver struct {
+	incoming   chan msgErr
+	manualAcks bool
+}
+
+func NewReceiver(opts ...ReceiverOption) (*Receiver, error) {
+	r := &Receiver{
+		incoming: make(chan msgErr),
+	}
+
+	if err := r.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+// MsgHandler implements stan.MsgHandler
+// This function is passed to the call to stan.Conn.Subscribe so that we can stream messages to be delivered
+// via Receive()
+func (r *Receiver) MsgHandler(msg *stan.Msg) {
+	m := (binding.Message)(NewMessage(msg))
+
+	if r.manualAcks {
+		m = binding.WithFinish(m, func(err error) {
+			if err == nil {
+				_ = msg.Ack()
+			}
+		})
+	}
+
+	r.incoming <- msgErr{msg: m}
+}
+
+// Receive implements Receiver.Receive
+// This should probably not be invoked directly by applications or library code, but instead invoked via
+// Protocol.Receive
+func (r *Receiver) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case msgErr, ok := <-r.incoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		return msgErr.msg, msgErr.err
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+func (r *Receiver) applyOptions(opts ...ReceiverOption) error {
+	for _, fn := range opts {
+		if err := fn(r); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type Consumer struct {
+	*Receiver
+
+	Conn               stan.Conn
+	Subject            string
+	Subscriber         Subscriber
+	UnsubscribeOnClose bool
+
+	subscriptionOptions []stan.SubscriptionOption
+	sub                 stan.Subscription
+	connOwned           bool                      // whether this consumer is responsible for closing the connection
+	appliedSubOpts      *stan.SubscriptionOptions // only used locally, stored to avoid recomputing
+}
+
+func NewConsumer(clusterID, clientID, subject string, stanOpts []stan.Option, opts ...ConsumerOption) (*Consumer, error) {
+	conn, err := stan.Connect(clusterID, clientID, stanOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewConsumerFromConn(conn, subject, opts...)
+}
+
+func NewConsumerFromConn(conn stan.Conn, subject string, opts ...ConsumerOption) (*Consumer, error) {
+	c := &Consumer{
+		Conn:    conn,
+		Subject: subject,
+	}
+
+	err := c.applyOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// OpenInbound implements Opener.OpenInbound.
+func (c *Consumer) OpenInbound(_ context.Context) error {
+	var err error
+	if c.sub != nil && c.sub.IsValid() {
+		// TODO: is this what we want?
+		return ErrSubscriptionAlreadyOpen
+	}
+
+	receiverOps, err := c.createReceiverOptions()
+	if err != nil {
+		return err
+	}
+
+	if c.Receiver, err = NewReceiver(receiverOps...); err != nil {
+		return err
+	}
+
+	c.sub, err = c.Subscriber.Subscribe(c.Conn, c.Subject, c.Receiver.MsgHandler, c.subscriptionOptions...)
```

Comment:
Should we wait for `ctx.Done()` and call `c.Close()`?

Reply:
In case of Kafka (and I think it's the same for http) `ctx.Done()` signals the stop of the receiver